### PR TITLE
20417 cookie config attributes from web xml

### PR DESCRIPTION
**Pull request by pmd1nh**
for #20417

SessionCookieConfig setting for attributes from the web.xml

### CODE REVIEW

**Comment by pmd1nh** [comment]
#build
        #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

**Review by zhang-ca — APPROVED** [review]
Looks good.

**Comment by pmd1nh** [comment]
results got delete even though it reported yesterday.  resubmit

#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

**Review by pnicolucci — APPROVED** [review]

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

**Comment by pmd1nh** [comment]
Updated test to cover precedence order - SCI setAttribute to override web.xml

#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat,com.ibm.ws.security.saml.sso_fat.config

**Comment by pmd1nh** [comment]
One failed but has been reported

junit.framework.AssertionFailedError: 2022-09-30-05:33:07:461 expected:<0> but was:<1>
	at com.ibm.ws.automation.test.HealingTest.testLibertyRunbook(HealingTest.java:213)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=289998

**Comment by pmd1nh** [comment]
@pnicolucci  @zhang-ca  this rework is ready for re-review.

**Comment by pmd1nh** [comment]
1. Change the logic to when the SessionCookieConfigImpl60 should be created.

    a. if there is a cookie-config element in web.xml
    OR
    b. right before the application start AND there still no SCC in the WebAppConfiguration. (Need to create at that time in case programmatically SCI add the cookie config during app startup). Also populated its value with the SessionManagerConfig's SCC.

2. Update the test to cover the precedent order SCI attribute > web.xml attribute

**Review by zhang-ca — APPROVED** [review]
Looks good. Thanks

File: dev/io.openliberty.webcontainer.servlet.6.0.internal_fat/test-applications/SessionCookieConfigSCI.war/resources/WEB-INF/web.xml [hunk 1]
```diff
@@ -10,6 +10,18 @@
             <max-age>2021</max-age>
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.0.internal_fat/test-applications/SessionCookieConfigSCI.war/resources/WEB-INF/web.xml:21]: Phu, can you add a copyright to this web.xml please.

**Review by pnicolucci — COMMENTED** [review]
Minor copyright update request to the web.xml. Also are there any other tests to ensure precedence programmatic->web.xml->server.xml  precedence is preserved (highest priority to lowest)?

File: dev/com.ibm.ws.session/src/com/ibm/ws/session/SessionManagerConfig.java [hunk 2]
```diff
@@ -1166,31 +1174,44 @@ public final void setSessionCookieSameSite(SameSiteCookie sameSite) {
...
             
-            if (scc.getComment() != null) {
-                this.setSessionCookieComment(scc.getComment());
+            if (TraceComponent.isAnyTracingEnabled() && LoggingUtil.SESSION_LOGGER_CORE.isLoggable(Level.FINER)) {
+                LoggingUtil.SESSION_LOGGER_CORE.log(Level.FINE, methodClassName + " updateCookieInfo , replace [" + cookieConfig +"] with ["+ scc + "] for this -> " + this);
+            }
+
+            //Servlet 6 update - webAppConfig SCC should populate this SMC. Any null/unset webAppConfig's SCC field is
+            //replaced with the current SMC's field before webAppConfig SCC replacing the SMC's SCC
+            //The webAppConfig SCC can be an instance of SCC 6 and may have addition attributes (already parsed from web.xml or set later by application via SCI)
```
Comment (pnicolucci) [inline dev/com.ibm.ws.session/src/com/ibm/ws/session/SessionManagerConfig.java]: may have `additional` attributes

File: dev/com.ibm.ws.session/src/com/ibm/ws/session/SessionManagerConfig.java [hunk 3]
```diff
@@ -1166,31 +1174,44 @@ public final void setSessionCookieSameSite(SameSiteCookie sameSite) {
...
+            //Also, a little later after this method is called, the SessionContextRegistryImpl will set both Session SMC and webAppConfig
+            //to the same SCC anyway (around line 318)
```
Comment (pnicolucci) [inline dev/com.ibm.ws.session/src/com/ibm/ws/session/SessionManagerConfig.java:1190]: Should you add a method name here etc as line number tend to change pretty frequently?

**Comment by pmd1nh** [comment]
updated minor comments

#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat

**Comment by pmd1nh** [comment]
#libby

**Pull request merged**